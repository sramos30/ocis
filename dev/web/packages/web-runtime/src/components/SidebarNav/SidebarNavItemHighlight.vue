<template>
  <span
    v-if="initial"
    id="highlighter"
    class="oc-background-primary-gradient"
    :data-initial-id="index"
  />
</template>
<script>
export default {
  props: {
    index: {
      type: Number,
      required: true
    },
    active: {
      type: Boolean,
      required: true
    }
  },
  data() {
    return {
      initial: false
    }
  },
  watch: {
    active(active) {
      if (!active) return
      this.animateHighlightPosition(this.index)
    }
  },
  created() {
    if (!this.active) return
    this.initial = true
  },
  methods: {
    animateHighlightPosition(target, durationSeconds = 0.2) {
      const initialId = this.getInitialId()
      const currentElement = this.getNavigationItem(initialId)
      const targetElement = this.getNavigationItem(target)
      const distanceAbs = this.getDistanceBetweenElements(currentElement, targetElement)
      const distance = target < initialId ? -distanceAbs : distanceAbs
      const style = this.getHighlightingElement().style
      style.setProperty('transition-duration', `${durationSeconds}s`)
      style.setProperty('transform', `translateY(${distance}px)`)
    },
    getNavigationItem(index) {
      return document.querySelectorAll(`[data-nav-id="${index}"]`)[0]
    },
    getPositionAtCenter(element) {
      const { top, left, width, height } = element.getBoundingClientRect()
      return {
        x: left + width / 2,
        y: top + height / 2
      }
    },
    getDistanceBetweenElements(a, b) {
      const aPosition = this.getPositionAtCenter(a)
      const bPosition = this.getPositionAtCenter(b)
      return Math.hypot(aPosition.x - bPosition.x, aPosition.y - bPosition.y)
    },
    getHighlightingElement() {
      return document.getElementById('highlighter')
    },
    getInitialId() {
      return this.getHighlightingElement().getAttribute('data-initial-id')
    }
  }
}
</script>

<style lang="scss">
#highlighter {
  border-radius: 5px;
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
  transition: transform 0.2s cubic-bezier(0.51, 0.06, 0.56, 1.37);
  z-index: -1;

  &::before {
    bottom: 0;
    box-shadow: 2px 0 6px rgba(0, 0, 0, 0.14);
    content: '';
    left: 0;
    position: absolute;
    top: 0;
    width: 40px;
  }
}
</style>
