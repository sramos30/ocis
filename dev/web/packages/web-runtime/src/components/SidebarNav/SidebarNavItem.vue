<template>
  <li class="oc-sidebar-nav-item oc-pb-xs oc-px-s" :aria-current="active ? 'page' : null">
    <oc-button
      v-oc-tooltip="toolTip"
      type="router-link"
      appearance="raw"
      :variation="active ? 'inverse' : 'passive'"
      :class="['oc-sidebar-nav-item-link', { active: active }]"
      :to="target"
      :data-nav-id="index"
      :data-nav-name="navName"
    >
      <span class="oc-flex">
        <oc-icon :name="icon" :fill-type="fillType" />
        <span class="oc-ml-m text" :class="{ 'text-invisible': collapsed }" v-text="name" />
      </span>
      <oc-tag v-if="tag" class="oc-py-rm" size="small">{{ tag }}</oc-tag>
      <sidebar-nav-item-highlight :index="index" :active="active" />
    </oc-button>
  </li>
</template>
<script>
import SidebarNavItemHighlight from './SidebarNavItemHighlight.vue'
import get from 'lodash-es/get'

export default {
  components: {
    SidebarNavItemHighlight
  },
  props: {
    name: {
      type: String,
      required: true
    },
    index: {
      type: Number,
      required: true
    },
    active: {
      type: Boolean,
      required: false,
      default: false
    },
    target: {
      type: String,
      required: false,
      default: null
    },
    icon: {
      type: String,
      required: true
    },
    fillType: {
      type: String,
      required: false,
      default: 'fill'
    },
    collapsed: {
      type: Boolean,
      required: false,
      default: false
    },
    tag: {
      type: String,
      required: false,
      default: null
    }
  },
  computed: {
    navName() {
      return get(this.$router?.resolve(this.target), 'route.name')
    },
    toolTip() {
      const value = this.collapsed
        ? this.$gettextInterpolate(this.$gettext('Navigate to %{ pageName } page'), {
            pageName: this.name
          })
        : ''
      return {
        content: value,
        placement: 'right',
        arrow: false
      }
    }
  }
}
</script>

<style lang="scss">
.oc-sidebar-nav-item-link {
  position: relative;
  align-items: center !important;
  display: flex !important;
  justify-content: space-between !important;
  padding: var(--oc-space-small) !important;
  border-radius: 5px;
  white-space: nowrap;
  user-select: none;

  .oc-tag {
    color: var(--oc-color-text-default);
    background-color: var(--oc-color-background-highlight);
  }
  .text {
    opacity: 1;
    transition: all 0s;
    transition-delay: 0.1s;
  }
  .text-invisible {
    opacity: 0 !important;
    transition: 0s;
  }

  &:hover:not(.active) {
    color: var(--oc-color-swatch-brand-hover) !important;
  }

  &:hover,
  &:focus {
    text-decoration: none !important;
  }
  &.active {
    cursor: default;
    overflow: hidden;
  }
}
</style>
