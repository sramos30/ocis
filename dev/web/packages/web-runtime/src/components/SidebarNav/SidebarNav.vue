<template>
  <div
    id="web-nav-sidebar"
    :class="{
      'oc-app-navigation-collapsed': navigation.closed,
      'oc-app-navigation-expanded': !navigation.closed
    }"
  >
    <oc-button
      appearance="raw"
      :class="toggleSidebarButtonClass"
      class="toggle-sidebar-button oc-py-s"
      :aria-label="$gettext('Toggle sidebar')"
      @click="toggleSidebarButtonClick"
    >
      <oc-icon size="large" fill-type="line" :name="toggleSidebarButtonIcon" />
    </oc-button>
    <nav class="oc-sidebar-nav oc-my-m oc-px-xs" :aria-label="$gettext('Sidebar navigation menu')">
      <oc-list>
        <sidebar-nav-item
          v-for="(link, index) in navItems"
          :key="link.route.path"
          :index="index"
          :target="link.route.path"
          :active="link.active"
          :icon="link.icon"
          :fill-type="link.fillType"
          :name="link.name"
          :collapsed="navigation.closed"
          :tag="link.tag"
        />
      </oc-list>
    </nav>
    <!-- @slot bottom content of the sidebar -->
    <slot name="bottom" />
  </div>
</template>

<script>
import { mapState, mapActions } from 'vuex'
import SidebarNavItem from './SidebarNavItem.vue'

export default {
  components: {
    SidebarNavItem
  },
  props: {
    navItems: {
      type: Array,
      required: true
    }
  },
  computed: {
    ...mapState(['navigation']),

    toggleSidebarButtonClass() {
      return this.navigation.closed
        ? 'toggle-sidebar-button-collapsed'
        : 'toggle-sidebar-button-expanded oc-pr-s'
    },

    toggleSidebarButtonIcon() {
      return this.navigation.closed ? 'arrow-drop-right' : 'arrow-drop-left'
    }
  },
  methods: {
    ...mapActions(['openNavigation', 'closeNavigation']),

    toggleSidebarButtonClick() {
      return this.navigation.closed ? this.openNavigation() : this.closeNavigation()
    }
  }
}
</script>

<style lang="scss">
#web-nav-sidebar {
  background-color: var(--oc-color-background-default);
  border-radius: 15px;
  box-shadow: 5px 0 25px rgba(0, 0, 0, 0.3);
  display: flex;
  flex-direction: column;
  overflow: hidden;
  transition: all 0.35s cubic-bezier(0.34, 0.11, 0, 1.12);
  z-index: 4;

  .toggle-sidebar-button {
    min-height: 3rem;
    transition: all 0.2s ease-out;
    &:hover {
      background-color: var(--oc-color-background-secondary) !important;
      overflow: hidden;
    }
  }
  .toggle-sidebar-button-expanded {
    justify-content: flex-end !important;
  }
}
.oc-app-navigation-expanded {
  min-width: 250px !important;
  max-width: 250px !important;
}
.oc-app-navigation-collapsed {
  min-width: 64px !important;
  max-width: 64px !important;
}
</style>
