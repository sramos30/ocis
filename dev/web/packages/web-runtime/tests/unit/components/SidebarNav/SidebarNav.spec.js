import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'

import SidebarNav from 'web-runtime/src/components/SidebarNav/SidebarNav.vue'

import stubs from '../../../../../../tests/unit/stubs'
import sidebarNavItemFixtures from '../../../../../../__fixtures__/sidebarNavItems'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(DesignSystem)

const slots = {
  bottom: '<span class="footer">Footer</span>'
}

describe('OcSidebarNav', () => {
  const wrapper = getWrapper()

  it('displays a bottom slot if given', () => {
    expect(wrapper.findAll('.footer').length).toBe(1)
  })

  it('renders navItems into a list', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('toggles into closed state upon button click', async () => {
    await wrapper.find('.toggle-sidebar-button').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper).toMatchSnapshot()
  })

  it('toggles back into open state upon button click', async () => {
    await wrapper.find('.toggle-sidebar-button').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper).toMatchSnapshot()
  })
})

function getWrapper() {
  return mount(SidebarNav, {
    store: new Vuex.Store({
      state: {
        navigation: {
          closed: false
        }
      },
      mutations: {
        SET_CLOSED(state, value) {
          state.navigation.closed = value
        }
      },
      actions: {
        openNavigation({ commit }) {
          commit('SET_CLOSED', false)
        },
        closeNavigation({ commit }) {
          commit('SET_CLOSED', true)
        }
      }
    }),
    localVue,
    stubs: {
      ...stubs,
      'sidebar-nav-item': true,
      'oc-button': false
    },
    slots,
    propsData: {
      navItems: sidebarNavItemFixtures
    }
  })
}
