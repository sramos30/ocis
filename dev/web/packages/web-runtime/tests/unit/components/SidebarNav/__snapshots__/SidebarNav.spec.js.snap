// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OcSidebarNav renders navItems into a list 1`] = `
<div id="web-nav-sidebar" class="oc-app-navigation-expanded"><button aria-label="Toggle sidebar" type="button" class="toggle-sidebar-button oc-py-s oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw toggle-sidebar-button-expanded oc-pr-s">
    <oc-icon-stub name="arrow-drop-left" filltype="line" accessiblelabel="" type="span" size="large" variation="passive" color=""></oc-icon-stub>
  </button>
  <nav aria-label="Sidebar navigation menu" class="oc-sidebar-nav oc-my-m oc-px-xs">
    <oc-list-stub>
      <sidebar-nav-item-stub name="Personal" index="0" active="true" target="/files/list/all" icon="folder" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shares" index="1" target="/files/list/shared-with-me" icon="share-forward" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Deleted files" index="2" target="/files/list/trash-bin" icon="delete" filltype="fill"></sidebar-nav-item-stub>
    </oc-list-stub>
  </nav> <span class="footer">Footer</span>
</div>
`;

exports[`OcSidebarNav toggles back into open state upon button click 1`] = `
<div id="web-nav-sidebar" class="oc-app-navigation-expanded"><button aria-label="Toggle sidebar" type="button" class="toggle-sidebar-button oc-py-s oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw toggle-sidebar-button-expanded oc-pr-s">
    <oc-icon-stub name="arrow-drop-left" filltype="line" accessiblelabel="" type="span" size="large" variation="passive" color=""></oc-icon-stub>
  </button>
  <nav aria-label="Sidebar navigation menu" class="oc-sidebar-nav oc-my-m oc-px-xs">
    <oc-list-stub>
      <sidebar-nav-item-stub name="Personal" index="0" active="true" target="/files/list/all" icon="folder" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shares" index="1" target="/files/list/shared-with-me" icon="share-forward" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Deleted files" index="2" target="/files/list/trash-bin" icon="delete" filltype="fill"></sidebar-nav-item-stub>
    </oc-list-stub>
  </nav> <span class="footer">Footer</span>
</div>
`;

exports[`OcSidebarNav toggles into closed state upon button click 1`] = `
<div id="web-nav-sidebar" class="oc-app-navigation-collapsed"><button aria-label="Toggle sidebar" type="button" class="toggle-sidebar-button oc-py-s oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw toggle-sidebar-button-collapsed">
    <oc-icon-stub name="arrow-drop-right" filltype="line" accessiblelabel="" type="span" size="large" variation="passive" color=""></oc-icon-stub>
  </button>
  <nav aria-label="Sidebar navigation menu" class="oc-sidebar-nav oc-my-m oc-px-xs">
    <oc-list-stub>
      <sidebar-nav-item-stub name="Personal" index="0" active="true" target="/files/list/all" icon="folder" filltype="fill" collapsed="true"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shares" index="1" target="/files/list/shared-with-me" icon="share-forward" filltype="fill" collapsed="true"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Deleted files" index="2" target="/files/list/trash-bin" icon="delete" filltype="fill" collapsed="true"></sidebar-nav-item-stub>
    </oc-list-stub>
  </nav> <span class="footer">Footer</span>
</div>
`;
